### PR TITLE
feat: plumb method closure through to code generation

### PR DESCRIPTION
**Pull request by paulbutcher**
Doesn't do code generation yet, but probably worth reviewing and merging?

### CODE REVIEW

**Comment by paulbutcher** [comment]
#4291

**Comment by magnus-madsen** [comment]
> Doesn't do code generation yet, but probably worth reviewing and merging?

Yes, lets do it.

**Comment by paulbutcher** [comment]
Although as you can see the tests have just failed. Not sure why: looking.

File: main/src/ca/uwaterloo/flix/language/phase/ClosureConv.scala [hunk 1]
```diff
@@ -279,7 +266,8 @@ object ClosureConv {
     case Expression.NewObject(clazz, tpe, purity, methods0, loc) =>
       val methods = methods0 map {
         case JvmMethod(ident, fparams, exp, retTpe, purity, loc) =>
-          JvmMethod(ident, fparams, visitExp(exp), retTpe, purity, loc)
+          val closure = createLambdaClosure(fparams, exp, retTpe, loc)
+          JvmMethod(ident, fparams, closure, retTpe, purity, loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ClosureConv.scala:270]: I think the type used here for `JvmMethod` might be wrong. See:

If we have `x -> x + 1`, a lambda, then it has type `Int32 -> Int32`.

If we write:

```
object Foo {
  def inc(x: Int32): Int32 = x + 1 
}
```
then the type of `exp` will be Int32, not `Int32 -> Int32.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/ClosureConv.scala:270]: Thus here we must build a function type from the type of the fparams and retType.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala [hunk 2]
```diff
@@ -255,7 +255,7 @@ object LiftedAst {
 
   case class Case(sym: Symbol.EnumSym, tag: Name.Tag, tpeDeprecated: Type, loc: SourceLocation)
 
-  case class JvmMethod(ident: Name.Ident, fparams: List[LiftedAst.FormalParam], retTpe: Type, purity: Purity, loc: SourceLocation)
+  case class JvmMethod(ident: Name.Ident, fparams: List[LiftedAst.FormalParam], closure: LiftedAst.Expression, retTpe: Type, purity: Purity, loc: SourceLocation)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala:258]: closure -> clo

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 3]
```diff
@@ -266,7 +266,7 @@ object ErasedAst {
 
   case class Case(sym: Symbol.EnumSym, tag: Name.Tag, tpeDeprecated: MonoType, loc: SourceLocation)
 
-  case class JvmMethod(ident: Name.Ident, fparams: List[ErasedAst.FormalParam], retTpe: MonoType, loc: SourceLocation)
+  case class JvmMethod(ident: Name.Ident, fparams: List[ErasedAst.FormalParam], closure: ErasedAst.Expression, retTpe: MonoType, loc: SourceLocation)
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:269]: On second thought, you will need the return type to generate the right method signature. So leave it in and keep it as the return type!

File: main/src/ca/uwaterloo/flix/language/phase/ClosureConv.scala (hunk 1 above)
Comment (paulbutcher) [reply main/src/ca/uwaterloo/flix/language/phase/ClosureConv.scala:270]: I think that this comment is now moot given that the closure type is now gone?

**Comment by paulbutcher** [comment]
Hmm, I tried to get clever with this pull request and rebase it onto the closure tidyup branch I created earlier. But this doesn't seem to have done quite what I hoped.

Hopefully this will sort itself out when the closure tidyup PR is merged and this is rebased against master? If not, I'll kill this and recreate.

**Comment by magnus-madsen** [comment]
I think it needs some kind of rebase.

**Comment by paulbutcher** [comment]
Right, think I've got it back into a sane state 👍

File: main/src/ca/uwaterloo/flix/language/ast/FinalAst.scala [hunk 4]
```diff
@@ -202,7 +202,7 @@ object FinalAst {
 
   case class Case(sym: Symbol.EnumSym, tag: Name.Tag, tpeDeprecated: MonoType, loc: SourceLocation)
 
-  case class JvmMethod(ident: Name.Ident, fparams: List[FinalAst.FormalParam], retTpe: MonoType, loc: SourceLocation)
+  case class JvmMethod(ident: Name.Ident, fparams: List[FinalAst.FormalParam], closure: FinalAst.Expression, retTpe: MonoType, loc: SourceLocation)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/FinalAst.scala:205]: closure -> clo

File: main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala [hunk 5]
```diff
@@ -256,7 +256,7 @@ object OccurrenceAst {
 
   case class Case(sym: Symbol.EnumSym, tag: Name.Tag, tpeDeprecated: Type, loc: SourceLocation)
 
-  case class JvmMethod(ident: Name.Ident, fparams: List[OccurrenceAst.FormalParam], retTpe: Type, purity: Purity, loc: SourceLocation)
+  case class JvmMethod(ident: Name.Ident, fparams: List[OccurrenceAst.FormalParam], closure: OccurrenceAst.Expression, retTpe: Type, purity: Purity, loc: SourceLocation)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala:259]: closure -> clo

File: main/src/ca/uwaterloo/flix/language/dbg/PrettyPrinter.scala [hunk 6]
```diff
@@ -498,10 +498,10 @@ object PrettyPrinter {
     }
 
     def fmtJvmMethod(method: JvmMethod, formatter: Formatter): String = method match {
-      case JvmMethod(ident, fparams, retTpe, purity, loc) =>
+      case JvmMethod(ident, fparams, closure, retTpe, purity, loc) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/dbg/PrettyPrinter.scala:501]: closure -> clo

File: main/src/ca/uwaterloo/flix/language/phase/ClosureConv.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/ClosureConv.scala:270]: I think so.

File: main/src/ca/uwaterloo/flix/language/phase/Finalize.scala [hunk 7]
```diff
@@ -514,9 +514,10 @@ object Finalize {
   }
 
   private def visitJvmMethod(method: LiftedAst.JvmMethod, m: TopLevel)(implicit flix: Flix): FinalAst.JvmMethod = method match {
-    case LiftedAst.JvmMethod(ident, fparams, retTpe, purity, loc) =>
+    case LiftedAst.JvmMethod(ident, fparams, closure, retTpe, purity, loc) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Finalize.scala:517]: closure -> clo

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala [hunk 8]
```diff
@@ -402,11 +402,12 @@ object Inliner {
 
     case OccurrenceAst.Expression.NewObject(clazz, tpe, purity, methods0, loc) =>
       val methods = methods0.map {
-        case OccurrenceAst.JvmMethod(ident, fparams, retTpe, purity, loc) =>
+        case OccurrenceAst.JvmMethod(ident, fparams, closure, retTpe, purity, loc) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:405]: closure -> clo

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala [hunk 9]
```diff
@@ -796,11 +797,12 @@ object Inliner {
...
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:800]: ditto

File: main/src/ca/uwaterloo/flix/language/phase/LambdaLift.scala [hunk 10]
```diff
@@ -376,9 +375,14 @@ object LambdaLift {
     }
 
     def visitJvmMethod(method: SimplifiedAst.JvmMethod): LiftedAst.JvmMethod = method match {
-      case SimplifiedAst.JvmMethod(ident, fparams0, _, retTpe, purity, loc) =>
+      case SimplifiedAst.JvmMethod(ident, fparams0, exp, retTpe, purity, loc) =>
         val fparams = fparams0 map visitFormalParam
-        LiftedAst.JvmMethod(ident, fparams, retTpe, purity, loc)
+        visitExp(exp) match {
+          case closure: LiftedAst.Expression.Closure =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/LambdaLift.scala:381]: This pattern match should no longer be required right? Now you can just use the result of visitExp directly?

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 11]
```diff
@@ -410,11 +410,12 @@ object OccurrenceAnalyzer {
 
     case Expression.NewObject(clazz, tpe, purity, methods, loc) =>
       val (ms, o1) = methods.map {
-        case LiftedAst.JvmMethod(ident, fparams, retTpe, purity, loc) => {
+        case LiftedAst.JvmMethod(ident, fparams, closure, retTpe, purity, loc) => {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:413]: ditto

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenAnonymousClasses.scala [hunk 12]
```diff
@@ -82,7 +82,7 @@ object GenAnonymousClasses {
     * Method
     */
   private def compileMethod(method: JvmMethod, visitor: ClassWriter)(implicit root: Root, flix: Flix): Unit = method match {
-    case JvmMethod(ident, fparams, tpe, loc) =>
+    case JvmMethod(ident, fparams, closure, tpe, loc) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenAnonymousClasses.scala:85]: ditto

File: main/src/ca/uwaterloo/flix/language/phase/jvm/JvmOps.scala [hunk 13]
```diff
@@ -1019,10 +1019,11 @@ object JvmOps {
 
       case Expression.NewObject(_, _, methods, _) => 
         methods.foldLeft(Set.empty[MonoType]) {
-          case (sacc, JvmMethod(_, fparams, retTpe, _)) =>
-            sacc ++ fparams.foldLeft(Set(retTpe)) {
-              case (acc, FormalParam(_, tpe)) => acc + tpe
-            }
+          case (sacc, JvmMethod(_, fparams, closure, retTpe, _)) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/JvmOps.scala:1022]: ditto

**Review by magnus-madsen — COMMENTED** [review]
Added lots of trivial comments and one important comment about a pattern match that is no longer needed.

**Comment by paulbutcher** [comment]
Thanks @magnus-madsen, and apologies for being lazy. Your previous comments were sufficient, and I should have had the sense to go through and find all the `closure` -> `clo` changes necessary. Next time, please don't feel that you have to go through and mark each instance, you can just tell me to stop being lazy.

**Comment by magnus-madsen** [comment]
> Thanks @magnus-madsen, and apologies for being lazy. Your previous comments were sufficient, and I should have had the sense to go through and find all the `closure` -> `clo` changes necessary. Next time, please don't feel that you have to go through and mark each instance, you can just tell me to stop being lazy.

Hah, no worries. I think I was half asleep :P

**Comment by magnus-madsen** [comment]
Merged. Thanks!

**Pull request merged**